### PR TITLE
Pcs stop before deploying, start after

### PR DESCRIPTION
**Pull request by dkurepa**
https://github.com/dotnet/arcade-services/issues/3202
This PR improves the PCS deployments by calling calling the `stop` endpoint, telling the service to finish it's job and stop working, and doing the deployment then.
If the deployment fails, we start the old revision again.
If it succeeds, we start the new one, while the old one remains stopped

### Release Note Category
- [ ] Feature changes/additions 
- [ ] Bug fixes
- [x] Internal Infrastructure Improvements

### Release Note Description
Stop the PCS before deployment, start it afterwards

### CODE REVIEW

File: eng/deployment/product-construction-service-deploy.ps1 [hunk 1]
```diff
@@ -42,39 +77,52 @@ else
...
+        Write-Host "Assigning label $inactiveLabel to the new revision"
+        # assign the label to the new revision
+        az containerapp revision label add --label $inactiveLabel --name $containerappName --resource-group $resourceGroupName --revision $newRevisionName | Out-Null
+
+        # transfer all traffic to the new revision
+        az containerapp ingress traffic set --name $containerappName --resource-group $resourceGroupName --label-weight "$inactiveLabel=100" | Out-Null
+        Write-Host "All traffic has been redirected to label $inactiveLabel"
+    }
+    else {
+        Write-Host "New revision is not running. Check revision $newRevisionName logs in the inactive revisions. Deactivating the new revision"
```
Comment (premun) [inline eng/deployment/product-construction-service-deploy.ps1:118]: This should be written as error?

File: eng/deployment/product-construction-service-deploy.ps1 [hunk 2]
```diff
@@ -42,39 +77,52 @@ else
...
-    Write-Host "All traffic has been redirected to label $inactiveLabel"
 }
-else {
-    Write-Host "New revision is not running. Check revision $newRevisionName logs in the inactive revisions. Deactivating the new revision"
-    az containerapp revision deactivate --revision $newRevisionName --name $containerappName --resource-group $resourceGroupName
-    exit 1
-}
\ No newline at end of file
+finally {
+    # Start the service. This either starts the new revision or the old one if the new one failed to start
```
Comment (premun) [inline eng/deployment/product-construction-service-deploy.ps1:124]: Does this get called even after `exit 1`?

File: eng/deployment/product-construction-service-deploy.ps1 [hunk 3]
```diff
@@ -42,39 +77,52 @@ else
...
+    Write-Host "Starting the product construction service"
+    $pcsStartUrl = $pcsUrl + "/status/start"
```
Comment (premun) [inline eng/deployment/product-construction-service-deploy.ps1:126]: We could use the status url here.

Or maybe better. Just define 3 variables on top of the file - status, start and stop URLs and use them through out the script?

File: eng/deployment/product-construction-service-deploy.ps1 (hunk 2 above)
Comment (dkurepa) [reply eng/deployment/product-construction-service-deploy.ps1:124]: it does, yes https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_try_catch_finally?view=powershell-7.4#:~:text=A%20finally%20block%20runs,within%20a%20catch%20block.

File: eng/deployment/product-construction-service-deploy.ps1 [hunk 4]
```diff
@@ -6,9 +6,44 @@ param(
...
+    [Parameter(Mandatory=$true)][string]$pcsUrl
 )
 
+function StopAndWait([string]$pcsUrl) {
+    try {
+        $pcsStopUrl = $pcsUrl + "/status/stop"
+        $stopResponse = Invoke-WebRequest -Uri $pcsStopUrl -Method Put
+
+        if ($stopResponse.StatusCode -ne 200) {
+            Write-Host "Service isn't responding to the stop request. Deploying the new revision without stopping the service."
```
[the remark this answers is not in the document]
Comment (dkurepa) [reply eng/deployment/product-construction-service-deploy.ps1]: Done

File: eng/deployment/product-construction-service-deploy.ps1 [hunk 5]
```diff
@@ -6,9 +6,44 @@ param(
...
+            return
+        }
+
+        # wait for the service to finish processing the current job
+        $sleep = $false
+        $pcsStatusUrl = $pcsUrl + "/status"
```
[the remark this answers is not in the document]
Comment (dkurepa) [reply eng/deployment/product-construction-service-deploy.ps1]: Done

File: eng/deployment/product-construction-service-deploy.ps1 [hunk 6]
```diff
@@ -6,9 +6,44 @@ param(
...
+        DO
+        {
+            if ($sleep -eq $true) 
+            {
+                Start-Sleep -Seconds 30
+            }
+            $pcsStateResponse = Invoke-WebRequest -Uri $pcsStatusUrl -Method Get
+            if ($pcsStateResponse.StatusCode -ne 200) {
+                Write-Host "Service isn't responding to the status request. Deploying the new revision without stopping the service."
```
[the remark this answers is not in the document]
Comment (dkurepa) [reply eng/deployment/product-construction-service-deploy.ps1]: Done

File: eng/deployment/product-construction-service-deploy.ps1 [hunk 7]
```diff
@@ -6,9 +6,44 @@ param(
...
+                return
+            }
+            Write-Host "Product Construction Service state: $($pcsStateResponse.Content)"
+            $sleep = $true
+        } While ($pcsStateResponse.Content -notmatch "Stopped")
+    }
+    catch {
+        Write-Host "An error occurred: $($_.Exception.Message).  Deploying the new revision without stopping the service."
```
[the remark this answers is not in the document]
Comment (dkurepa) [reply eng/deployment/product-construction-service-deploy.ps1]: Done

File: eng/deployment/product-construction-service-deploy.ps1 (hunk 1 above)
Comment (dkurepa) [reply eng/deployment/product-construction-service-deploy.ps1]: Think so, done

File: eng/deployment/product-construction-service-deploy.ps1 (hunk 3 above)
Comment (dkurepa) [reply eng/deployment/product-construction-service-deploy.ps1]: Done

**Review by premun — APPROVED** [review]

**Comment by MilenaHristova** [comment]
Rolled out with https://github.com/dotnet/arcade-services/issues/3288

**Pull request merged**